core/crypto/eddsa/blinded25519.go

```diff
@@ -0,0 +1,169 @@
+	// secret/private scalar directly as expandedSecretKey
+	// instead of deriving from hash each time.
+	// It is a bit unfortunate that this needs to be here...
+	blindedSecretKey := [64]byte{}
+	copy(blindedSecretKey[:], privateKey.blinded[:32])
+	expandedSecretKey := new(edwards25519.Scalar)
+	expandedSecretKey.SetUniformBytes(blindedSecretKey[:])
+
+	// secret_salt <- sha512(private)
+	digest1 := sha512.Sum512(privateKey.blinded[:32])
+
+	// very important that nonce is never repeated for different messages;
+	// ensured here by deriving nonce from a hash of the message.
+	// the second appendix of secret_salt below is not part of the spec:
+	// nonce := sha512(secret_salt[32:64] || message || secret_salt[33:64])
+	h := sha512.New()
+	h.Write(digest1[32:])
+	h.Write(message)
+	h.Write(digest1[33:])
+	messageDigest := h.Sum(nil)
+
+	// messageDigestReduced <- (nonce mod L)
+	mdReduced, _ := new(edwards25519.Scalar).SetUniformBytes(messageDigest[:64])
+
+	// R <- messageDigestReduced * B
+	var encodedR [32]byte
+	copy(encodedR[:], new(edwards25519.Point).ScalarBaseMult(mdReduced).Bytes())
+
+	// hramDigestReduced := sha512(R || public || message) mod L
+	h.Reset()
+	h.Write(encodedR[:])
+	h.Write(privateKey.blinded[32:])
+	h.Write(message)
+	hramDigest := h.Sum(nil)
+	hramDigestReduced, _ := new(edwards25519.Scalar).SetUniformBytes(hramDigest[:])
+	//     s := (hRamDigestReduced*secret + messageDigestReduced) mod L
+	// <-> s := (   (sha512(R || public || message) mod L)*secret
+	//            + (nonce mod L)
+	//          ) mod L
+	s_new := new(edwards25519.Scalar).MultiplyAdd(hramDigestReduced,
+		expandedSecretKey,
+		mdReduced)
+
+	copy(signature[:], encodedR[:])
+	copy(signature[32:], s_new.Bytes())
+	return signature
+}
+
+func hash_factor_inplace(factor *[32]byte) {
+	h := sha512.Sum512_256(factor[:])
+	copy(factor[:], h[:])
+	// needs to be clamped for scalar multiplication,
+	// but Scalar.SetBytesWithClamping takes care of that.
+}
+
+func (secret *PrivateKey) Blind(factor [32]byte) *BlindedPrivateKey {
+	hash_factor_inplace(&factor)
+	factor_sc, err := new(edwards25519.Scalar).SetBytesWithClamping(factor[:])
+	if err != nil {
+		// This only happens if factor is not 32 bytes.
+		panic(err)
+	}
+
+	// Here digest is the actual secret key derived from the seed:
+	digest := sha512.Sum512(secret.Bytes()[:32])
+	bb_sc, err := new(edwards25519.Scalar).SetBytesWithClamping(digest[:32])
+
+	// (ab + c) mod l = ScMulAdd(out, a, b, c)
+	// a := factor
+	// b := secret
+	// c := 0
+	// the (c*B) multiplication here is unfortunate but that was the easiest API
+	// to use in edwards25519, so...
+	// oo <- factor * bb + 0*B
+	oo_sc := new(edwards25519.Scalar).Multiply(factor_sc, bb_sc)
+
+	newsec := make(ed25519.PrivateKey, 64)
+	copy(newsec[:32], oo_sc.Bytes())
+
+	// calculates and sets the public key corresponding to the
+	// private scalar (used when we have derived a new private scalar)
+	newsec_a := [64]byte{}
```

Comment:
https://github.com/FiloSottile/edwards25519/blob/main/scalar.go#L93
already does a copy so this is not necessary

ie just do
```newsec_scalar, _ := new(edwards25519.Scalar).SetUniformBytes(oo_src.Bytes()[:32])```

Reply:
For some reason this proposed change segfaults `go test`.
I don't understand why, since I agree with your interpretation that it should copy it.
Maybe we are using an older version of the library?